beacon-chain/rpc/prysm/v1alpha1/validator/proposer_bellatrix.go

```diff
@@ -50,47 +53,52 @@ var emptyTransactionsRoot = [32]byte{127, 254, 36, 30, 166, 1, 135, 253, 176, 24
 const blockBuilderTimeout = 1 * time.Second
 
 // Sets the execution data for the block. Execution data can come from local EL client or remote builder depends on validator registration and circuit breaker conditions.
-func setExecutionData(ctx context.Context, blk interfaces.SignedBeaconBlock, localPayload, builderPayload interfaces.ExecutionData, builderKzgCommitments [][]byte, builderBoostFactor uint64) error {
+func setExecutionData(ctx context.Context, blk interfaces.SignedBeaconBlock, local *blocks.GetPayloadResponse, bid builder.Bid, builderBoostFactor primitives.Gwei) (primitives.Wei, *enginev1.BlobsBundle, error) {
 	_, span := trace.StartSpan(ctx, "ProposerServer.setExecutionData")
 	defer span.End()
 
 	slot := blk.Block().Slot()
 	if slots.ToEpoch(slot) < params.BeaconConfig().BellatrixForkEpoch {
-		return nil
+		return primitives.ZeroWei, nil, nil
 	}
 
-	if localPayload == nil {
-		return errors.New("local payload is nil")
+	if local == nil {
+		return primitives.ZeroWei, nil, errors.New("local payload is nil")
 	}
 
 	// Use local payload if builder payload is nil.
-	if builderPayload == nil {
-		return setLocalExecution(blk, localPayload)
+	if bid == nil {
+		return local.Bid, local.BlobsBundle, setLocalExecution(blk, local)
 	}
 
-	switch {
-	case blk.Version() >= version.Capella:
-		// Compare payload values between local and builder. Default to the local value if it is higher.
-		localValueGwei, err := localPayload.ValueInGwei()
-		if err != nil {
-			return errors.Wrap(err, "failed to get local payload value")
-		}
-		builderValueGwei, err := builderPayload.ValueInGwei()
+	var builderKzgCommitments [][]byte
+	builderPayload, err := bid.Header()
+	if err != nil {
+		log.WithError(err).Warn("Proposer: failed to retrieve header from BuilderBid")
+		return local.Bid, local.BlobsBundle, setLocalExecution(blk, local)
+	}
+	if bid.Version() >= version.Deneb {
+		builderKzgCommitments, err = bid.BlobKzgCommitments()
 		if err != nil {
-			log.WithError(err).Warn("Proposer: failed to get builder payload value") // Default to local if can't get builder value.
-			return setLocalExecution(blk, localPayload)
+			log.WithError(err).Warn("Proposer: failed to retrieve kzg commitments from BuilderBid")
 		}
+	}
 
-		withdrawalsMatched, err := matchingWithdrawalsRoot(localPayload, builderPayload)
+	switch {
+	case blk.Version() >= version.Capella:
+		withdrawalsMatched, err := matchingWithdrawalsRoot(local.ExecutionData, builderPayload)
 		if err != nil {
 			tracing.AnnotateError(span, err)
 			log.WithError(err).Warn("Proposer: failed to match withdrawals root")
-			return setLocalExecution(blk, localPayload)
+			return local.Bid, local.BlobsBundle, setLocalExecution(blk, local)
 		}
 
+		// Compare payload values between local and builder. Default to the local value if it is higher.
+		localValueGwei := primitives.WeiToGwei(local.Bid)
+		builderValueGwei := primitives.WeiToGwei(bid.WeiValue())
```

Comment:
should we add a function on bid to return GweiValue instead? not sure which is more readable

Reply:
I only added the WeiValue method because it encapsulates the tricky reversing logic for accessing a value that is directly stored by the type. IMO the gwei conversion is pretty trivial and doesn't warrant another method.

Reply:
Also thinking about modifying the `Value` method to just be the `WeiValue` method since I don't think we want to expose the little endian version, but I'll need to double check usage.